### PR TITLE
add batch EI acquisition function

### LINKED ISSUE #41: MC-based batch-EI
A Monte-Carlo implementation of EI that works for a single query point or a batch of query points. The returned EI is always a scalar. Compared to analytical EI, the parameter eta is optional.

This acquisition fct uses the reparametrisation trick.

### PR DESCRIPTION
**Pull request by joelberkeley**
Adds the batch expected improvement acquisition function

### CODE REVIEW

File: trieste/acquisition/function.py [hunk 1]
```diff
@@ -625,3 +666,59 @@ def sample(self, at: TensorType, *, jitter: float = DEFAULTS.JITTER) -> tf.Tenso
...
+        )
+
+        eta = tf.reduce_min(mean, axis=0)
+        sampler = BatchReparametrizationSampler(self._sample_size, model)
+
+        def batch_ei(at: TensorType) -> TensorType:
+            samples = tf.squeeze(sampler.sample(at, jitter=self._jitter), axis=-1)  # [..., S, B]
+            min_sample_per_batch = tf.reduce_min(samples, axis=-1)  # [..., S]
+            batch_improvement = tf.maximum(eta - min_sample_per_batch, 0.0)  # [..., S]
+            return tf.reduce_mean(batch_improvement, axis=-1, keepdims=True)  # [..., 1]
```
Comment (joelberkeley) [inline trieste/acquisition/function.py:727]: you might want to double check this, as I've reordered it to be faster (found samples min before maximum)

Comment (vpicheny) [reply trieste/acquisition/function.py:727]: LGTM

File: tests/unit/acquisition/test_function.py [hunk 2]
```diff
@@ -546,20 +584,76 @@ def test_batch_reparametrization_sampler_samples_are_distinct_for_new_instances(
...
+
+@random_seed
+def test_batch_monte_carlo_expected_improvement() -> None:
+    xs = tf.random.uniform([3, 5, 7, 2], dtype=tf.float64)
+    model = QuadraticMeanAndRBFKernel()
+
+    mean, cov = model.predict_joint(xs)
+    mvn = tfp.distributions.MultivariateNormalFullCovariance(tf.linalg.matrix_transpose(mean), cov)
+    mvn_samples = mvn.sample(10_000)
+    expected = tf.reduce_mean(tf.reduce_max(tf.maximum(0.09 - mvn_samples, 0.0), axis=-1), axis=0)
```
Comment (vpicheny) [inline tests/unit/acquisition/test_function.py]: are we sure that eta=0.09 for acq?

Comment (vpicheny) [reply tests/unit/acquisition/test_function.py]: I would rather use `eta = tf.reduce_min(mean, axis=0)` instead of the hard-coded 0.09 above

**Review by vpicheny — APPROVED** [review]
looks good!

File: tests/unit/acquisition/test_function.py (hunk 2 above)
Comment (joelberkeley) [reply tests/unit/acquisition/test_function.py]: `QuadraticMeanAndRBFKernel` is a `GaussianProcess`, which outputs precisely the mean you tell it to. In this case, that's the quadratic function, so 0.3 ** 2

Comment (joelberkeley) [reply tests/unit/acquisition/test_function.py]: I could make that clearer by changing the 0.09 in the data below to have some noise on it, to make it clear they're not the same thing?

Comment (vpicheny) [reply tests/unit/acquisition/test_function.py]: Don't add noise! this is fine. I just find it hard to follow, there's no visible connexion between the model mean and 0.09. 
I may just not be used to the way tests ought to be written

Comment (joelberkeley) [reply tests/unit/acquisition/test_function.py]: i could
```
min_predictive_mean_at_known_points = 0.09
... (min_predictive_mean_at_known_points - mvn_samples) ...
```

Comment (vpicheny) [reply tests/unit/acquisition/test_function.py]: I like that
